reorganize & cleanup

ngam opened the pull request:
Closes #67 
Closes #14 
Closes #140 
Closes #87 
Closes #160 
Closes #138 
Closes #76 

TODO:
- [x] isolate julia inside conda
- [x] sync julia project with conda env
- [x] stack julia envs

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

ngam commented:
Supersedes #140

ngam commented:
@mkitti feel free to review and merge. Or let me know if there are other cleanup/organization things we should deal with first. We should also move this to the other branch (v1.6.5) once we settle on a procedure. 

Should we expand the tests above?

ngam commented:
@conda-forge-admin, please rerender

ngam commented:
These pass just fine locally: `# all except stdlib, add "opaque_closure", "filesystem" with 1.7`

ngam commented:
Okay, all look good --- will test the stdlib as well soon

mkitti commented:
Do we have parallelization in place yet?

ngam commented:
> Do we have parallelization in place yet?

Not yet. I have enabled all tests in the last commit --- we will see how far this will go (my local machine exited with OOM). Then, the last thing after that will be to set `ncores=ceil(Int, Sys.CPU_THREADS` instead of the default (?) `ncores=ceil(Int, Sys.CPU_THREADS / 2`

ngam commented:
Okay, I think this is it for this PR. The full test suite fails in random places, so we reverting back to the original list except adding new ones for v1.7 and increasing threads.

ngam commented:
@mkitti review/merge when you have a chance. Increasing threads makes the linux_64_ build finish in 50 mins as opposed to 75 mins.

mkitti commented:
I'm thinking of making a run at either setting a conda environment specific `JULIA_DEPOT_PATH` or at least setting `JULIA_PROJECT` ala #87 .

I wonder if downstream users have an opinion on this, @valeriupredoi ?

See also https://github.com/ESMValGroup/ESMValTool/issues/2419

ngam commented:
Two questions:

1. Do you want to do that part of this PR or another PR? We could squeeze it here for sure
2. As only a beginner Julia user, could you explain to me which option would basically make julia a sub env of the conda env? I believe from a zoomed out view, I think this should be our first default --- set up julia (and everything associated with it) inside the conda env and have julia not interact too much with the outside system (especially, say, another installation of julia somewhere)

mkitti commented:
That would be JULIA_DEPOT_PATH environment variable. That controls the DEPOT_PATH variable in Julia: https://docs.julialang.org/en/v1/base/constants/#Base.DEPOT_PATH

Essentially that would change the location of ~/.julia .

ngam commented:
Okay, so if we want to move `~/.julia` to `$PREFIX/etc/.julia`, do we just set `JULIA_DEPOT_PATH` to be `$PREFIX/etc/` via `export JULIA_DEPOT_PATH="$PREFIX/etc/:$JULIA_DEPOT_PATH"`?

Then, does that take care of unsetting `/usr/local/share/julia` and `/usr/share/julia`?

ngam commented:
Also, do you have preference if it is
`export JULIA_DEPOT_PATH="$CONDA_PREFIX/share/julia/site:$JULIA_DEPOT_PATH"`
vs
`export JULIA_DEPOT_PATH="$CONDA_PREFIX/etc/julia:$JULIA_DEPOT_PATH"` or
`export JULIA_DEPOT_PATH="$CONDA_PREFIX/etc/.julia:$JULIA_DEPOT_PATH"`

ngam commented:
(feel free to edit my branch btw as you see fit)

mkitti commented:
If `JULIA_DEPOT_PATH` is undefined, then you will get `$CONDA_PREFIX/share/julia/site:`. Julia will interpret the trailing `:` as meaning to append the default `DEPOT_PATH`. This could be useful.

```
(base) $ JULIA_DEPOT_PATH="" julia -E "DEPOT_PATH"
String[]
(base) $ JULIA_DEPOT_PATH=":" julia -E "DEPOT_PATH"
["/home/mkitti/.julia", "/home/mkitti/src/julia-1.6.0/local/share/julia", "/home/mkitti/src/julia-1.6.0/share/julia"]
(base) $ JULIA_DEPOT_PATH="$CONDA_PREFIX/share/julia/site:" julia -E "DEPOT_PATH"
["/home/mkitti/anaconda3/share/julia/site", "/home/mkitti/.julia", "/home/mkitti/src/julia-1.6.0/local/share/julia", "/home/mkitti/src/julia-1.6.0/share/julia"]
(base) $ JULIA_DEPOT_PATH="$CONDA_PREFIX/share/julia/site" julia -E "DEPOT_PATH"
["/home/mkitti/anaconda3/share/julia/site"]
```

mkitti commented:
Do we need something like https://github.com/conda-forge/openjdk-feedstock/blob/57916550977ace39ff53885556d8ee5a30bf254a/recipe/build.sh#L181-L187 ?

Then we have activate.sh
```
export JULIA_DEPOT_PATH_BACKUP=${JULIA_DEPOT_PATH:-}
export JULIA_PROJECT_BACKUP=${JULIA_PROJECT:-}
export JULIA_DEPOT_PATH="$CONDA_PREFIX/share/julia/site:$JULIA_DEPOT_PATH"
export JULIA_PROJECT="$CONDA_PREFIX/share/julia/site/environments/$CONDA_DEFAULT_ENV"
```

and deactivate.sh:
```
export JULIA_DEPOT_PATH=$JULIA_DEPOT_PATH_BACKUP
export JULIA_PROJECT=$JULIA_PROJECT_BACKUP
unset JULIA_DEPOT_PATH_BACKUP
unset JULIA_PROJECT_BACKUP

if [ -z $JULIA_DEPOT_PATH ]; then
    unset JULIA_DEPOT_PATH
fi
if [ -z $JULIA_PROJECT ]; then
    unset JULIA_PROJECT
fi
```

maybe we should just merge #87 into here and see what happens?

mkitti commented:
If I recall there were some minor test failures. Perhaps we just need to patch the tests?

ngam commented:
> If I recall there were some minor test failures. Perhaps we just need to patch the tests?

We will see. This latest merge from mjohnson541:keep-julia-local-to-conda-env shouldn't introduce anything substantial in terms of tests, but we will see soon.

ngam commented:
We have some tests failing...

ngam commented:
@mkitti, I believe the tests fail because of this: 
```
export JULIA_PROJECT="$CONDA_PREFIX/share/julia/site/environments/$CONDA_DEFAULT_ENV"
```

Why are we hardwiring an environment here? Doesn't julia start with its own default env?

ngam commented:
The other test fails because it basically points to home:

```
expanded == readchomp(setenv(`$exename -e 'println(Base.active_project())'`, "JULIA_PROJECT" => "@foo", "HOME" => homedir()))
```

mkitti commented:
Julia starts with its own major version specific environment. In this case, I think we want a project specific environment which may be independent of major version.

mkitti commented:
I agree the test is flawed. We should patch it and upstream it.

mkitti commented:
Why are we using $CONDA_DEFAULT_ENV above?

ngam commented:
> CONDA_DEFAULT_ENV

I think the original idea was to superficially sync the conda and julia envs. 

> I agree the test is flawed. We should patch it and upstream it.

I am tinkering with it now, I will try to figure out whatever julia's equivalent of $CONDA_PREFIX is and set that instead of home. However, I think the test is fundamentally broken, but I can be reading it wrong...

mkitti commented:
I just misunderstood the meaning of `$CONDA_DEFAULT_ENV`. I guess that just means the current conda environment...

mkitti commented:
I think we just need to pass `JULIA_DEPOT_PATH` through:

```julia
expanded == readchomp(setenv(`$exename -e 'println(Base.active_project())'`, "JULIA_PROJECT" => "@foo", "HOME" => homedir(), "JULIA_DEPOT_PATH" => ENV["JULIA_DEPOT_PATH"]))
```

ngam commented:
Yes! I will commit soon

mkitti commented:
> Yes! I will commit soon

Good. I'm getting myself a bit confused. It's possible this is fixed upstream.

mkitti commented:
About ILP64, see https://julialang.org/blog/2021/11/julia-1.7-highlights/#libblastrampoline_mkljl

mkitti commented:
Let's see what comes out of https://github.com/JuliaLang/julia/pull/43560 first, but this is looking good.

mkitti commented:
@conda-forge-admin, please restart ci

ngam commented:
> Let's see what comes out of [JuliaLang/julia#43560](https://github.com/JuliaLang/julia/pull/43560) first, but this is looking good.

The `addenv` alone won't cut it here because we are not running all tests. I mean --- I am sorry, that's a crazy hot garbage going up upstream in terms of testing. I will refix the my patches later today and we should merge.

I stand by my suggested edit here https://github.com/JuliaLang/julia/pull/43560#discussion_r775324245: adding ` "JULIA_DEPOT_PATH" => "$(Base.DEPOT_PATH[1])"` at the end of this call to ensure the correct point. 

And no, we don't need a `JULIA_PROJECT` --- we should just let julia decide whatever their default will be (v1.7) and the user can do whatever they want with that.

ngam commented:
@conda-forge-admin, please rerender

mkitti commented:
The reason we should define `JULIA_PROJECT` is that from the Julia perspective all the default environments will look like the generic base environment. The user will have no clue that they are actually in an environment that is encapsulated within their conda environment until they really interrogate it.

mkitti commented:
Running External
  Programs

ngam commented:
I see, but we never do that with anything else in conda. We also have the release tagline.

My understanding is that julia from conda should pretty much look identical as julia proper in a standalone installation. The only difference we are trying to achieve here is to _isolate_ julia (at least initially) within a conda environment. The other big difference is that we are relying on the system (read conda) to source some of the important dependencies instead of building them. Neither of these two will necessitate we predefine a specific env for the user. 

If you insist on presetting it, we will need to fix two tests or so. It's likely some tests are hardcoded upstream that need to be relaxed. Again though, I just don't think this is needed or will make a difference beyond making a mess of this solution. Since we haven't had any community engagement, you'll be the one to decide this 😉 so let me know what you want

mkitti commented:
@conda-forge-admin, please restart ci

ngam commented:
> @conda-forge-admin, please restart ci

Won't rereun the tests because with your latest addition of `addenv` it passed, so it won't repeat it unless something else changes.

ngam commented:
Oh, something may be wrong with the CI actually --- see: https://github.com/conda-forge/julia-feedstock/pull/159

ngam commented:
I can test the addition of JULIA_PROJECT locally. Should I focus on that?

mkitti commented:
The easiest way around the tests would probably be to do something like
```
julia --project -e 'delete!(ENV,"JULIA_PROJECT"); Base.runtests( ... )`
```

The `--project` will set the project to the default project, overriding the environmental variable. The `delete!(ENV,"JULIA_PROJECT")` will clear the environmental variable, so subsequent sub-processes do not inherit it.

ngam commented:
> The easiest way around the tests would probably be to do something like
> 
> ```
> julia --project -e 'delete!(ENV,"JULIA_PROJECT"); Base.runtests( ... )`
> ```
> 
> The `--project` will set the project to the default project, overriding the environmental variable. The `delete!(ENV,"JULIA_PROJECT")` will clear the environmental variable, so subsequent sub-processes do not inherit it.

But see, we can just drop these tests then. We should be testing for our particular case, where we have a new path and all, so that's why I didn't like the PR fix upstream because it still circumvents the legit test by going home again --- I would rather we tweak the test to address our own use case rather than circumventing them. The tests are there for a reason after all...

mkitti commented:
Well, we can try to patch it again, but the game is similar to the other issue.

The first test that fails is
https://github.com/JuliaLang/julia/blob/ac5cc99908d463582e66db3368b9b48fae1e2525/test/loading.jl#L671-L684
```julia
mktempdir() do dir
    vdir = Base.DEFAULT_LOAD_PATH[2]
    vdir = replace(vdir, "#" => VERSION.major, count = 1)
    vdir = replace(vdir, "#" => VERSION.minor, count = 1)
    vdir = replace(vdir, "#" => VERSION.patch, count = 1)
    vdir = vdir[2:end] # remove @
    vpath = joinpath(dir, "environments", vdir)
    mkpath(vpath)
    withenv("JULIA_DEPOT_PATH" => dir) do
        script = "@assert startswith(Base.active_project(), $(repr(vpath)))"
        @test success(`$(Base.julia_cmd()) --startup-file=no -e $(script)`)
    end
end
```

This could probably be modified such that `withenv` includes `JULIA_PROJECT`.
```julia
        mktempdir() do dir
           vdir = Base.DEFAULT_LOAD_PATH[2]
           vdir = replace(vdir, "#" => VERSION.major, count = 1)
           vdir = replace(vdir, "#" => VERSION.minor, count = 1)
           vdir = replace(vdir, "#" => VERSION.patch, count = 1)
           vdir = vdir[2:end] # remove @
           vpath = joinpath(dir, "environments", vdir)
           mkpath(vpath)
           withenv("JULIA_DEPOT_PATH" => dir, "JULIA_PROJECT" => "") do
               script = "@assert startswith(Base.active_project(), $(repr(vpath)))"
               @test success(`$(Base.julia_cmd()) --startup-file=no -e $(script)`)
           end
       end
```

ngam commented:
@mkitti let's shelve predefining a project for now. We can revisit in another round. I have serious reservations about the logic of adding it.

mkitti commented:
Once we merge this, I would not want to mess with the environment variables again until Julia 1.8.

Looking at my own `~/.julia`, which is 48 GB at the moment, I would much prefer to have a distinct `JULIA_PROJECT` per conda environment rather than an entirely new `JULIA_DEPOT_PATH` per conda environment. Having a `JULIA_DEPOT_PATH` per conda environment will just end up being too duplicative, especially in terms of artifacts. It will also make working with `dev`ed packages a bit troublesome.

There are some good reasons to have a distinct `JULIA_DEPOT_PATH` per conda environment at the moment (e.g. [PyCall.jl](https://github.com/JuliaPy/PyCall.jl)), but those are issues that are better resolved elsewhere in the long term.

I'm testing a modified `JULIA_PROJECT` and `JULIA_LOAD_PATH` in #160. What I think we should do is bring a commented out version of that to the activate.sh and deactivate.sh here, so that individual users can select that configuration if desired.

ngam commented:
@mkitti sounds good, this is definitely at the edge of comfort --- I am not familiar with how this works in practice. I will try to offer my zoomed out perspective in the other PR; specifically, there are few concerns I'd us to address (ideally add tests for) before we merge.

ngam commented:
Also, one more thing: How does this impact creating _additional_ julia envs inside a specific conda env?

> I would much prefer to have a distinct JULIA_PROJECT per conda environment rather than an entirely new JULIA_DEPOT_PATH per conda environment

mkitti commented:
The standard way to develop *packages* in Julia is to checkout their git trees in `$JULIA_DEPOT_PATH/dev`. For example, on this Linux box, I have the following packages in development.

```julia
$ ls ~/.julia/dev
BioformatsLoader  HDF5                  JLD       OpenCL        Spark            TaskWorkers
Demo              ImageTransformations  JProxies  RootTaskREPL  SpecialMatrices  Xnumber
FunctionRunner    Interpolations        Matcha    SHbundle      Sugar
GR                JavaCall              Napari    SnoopCompile  Taro
```

Users are highly encouraged to write packages rather than scripts. Packages are just Julia modules that are contained with Julia project environments. That is each package has their own Project.toml and Manifest.toml, although usually only Project.toml is distributed for libraries. One of the reasons packages are encouraged is that they are the basic precompilable unit. This precompilation is put in `$JULIA_DEPOT_PATH/compiled`.

As I mentioned above, each package is a Julia project. I can activate a package's project.

ngam commented:
> I'm testing a modified `JULIA_PROJECT` and `JULIA_LOAD_PATH` in #160. What I think we should do is bring a commented out version of that to the activate.sh and deactivate.sh here, so that individual users can select that configuration if desired.

@mkitti, I will leave this up to you. Please review, edit, and merge as you find most fit (and don't forget the LTS PR #159). If something goes wrong, we will be able to initiate another round of edits. But I trust this is pretty good development. We will also be hearing from people.

mkitti commented:
I think we should finish this one before addressing 1.6 LTS. I'm reluctant to making a breaking change to a patch release per semantic versioning and changing the JULIA_DEPOT_PATH will break setups.

Depots also stack in Julia, but I do not have a lot of experience with that.

ngam commented:
> I think we should finish this one before addressing 1.6 LTS. I'm reluctant to making a breaking change to a patch release per semantic versioning and changing the JULIA_DEPOT_PATH will break setups.

Do you want to keep this open until we get another minor update then?

> Once we merge this, I would not want to mess with the environment variables again until Julia 1.8.

What's their rough dev cycle upstream? When would you expect 1.7.2 and 1.8.0 to arrive?

ngam commented:
Do we have any other lingering issues that we should address btw? I saw some issues in this repo about packages but in my short testing, everything is working just fine. Basically, we have a complete julia 1.7.1 here that is equivalent to standalone installation, right?

(I am not talking about OS builds --- windows, arm, etc. will be a whole other hassle, that I am happy to work through in the next few weeks)

mkitti commented:
1.7.2 is probably weeks to a few months away. 1.7.0 and 1.7.1 were separated by three weeks.

1.8.0 is likely 6 months to a year away.

mkitti commented on recipe/build.sh:
```diff
@@ -6,6 +6,9 @@ export LIBRARY_PATH=${PREFIX}/lib
 export CMAKE_PREFIX_PATH=${PREFIX}
 export PATH="${PREFIX}/bin:${PATH}"
 
+#set JULIA_DEPOT_PATH in conda env
+export JULIA_DEPOT_PATH="${PREFIX}/share/julia/site:$JULIA_DEPOT_PATH" 
+
```
This should be removed.

ngam replied:
Done.

mkitti reviewed (changes requested):
Why are we keeping the old juliarc.jl and pre-unlink.sh around?

Is there any functionality in pre-unlink.sh that should be replicate?

mkitti commented:
I'm good to merge this if you are, @ngam .

ngam commented:
Should we address #161 or wait?

ngam commented:
@conda-forge-admin, please rerender

zklaus commented:
It's great to see the rapid improvements coming in here! Big steps forward on many fronts.

But if you want a bit more community engagement you might want to consider slowing down just a little bit. This PR, which closes 7 issues, some of which have been around for a long time and have been opened and worked upon by other people, took all of 5 days from creation to merging. Over the Christmas days.
Still, I don't want to dampen the enthusiasm; it really is great to see this move forward.


For now, I have two concerns/comments:

The issue of separate depots for separate environments is really all about binary compatibility. Basically, anything that has been compiled for one environment cannot reliably be used with another environment because linked libraries may have changed in an ABI incompatible way (cf the recent libunwind problems, but I would expect this with a number of libraries, e.g. for I/O of special file formats like netCDF) and the compilers themselves may be different. If compiled bits are shared, this will lead to surprising segfaults in the future. This kind of problem is not likely to show up in tests that use similar environments created at roughly the same time, but is virtually guaranteed if very different environments with library overlap are used, particularly when those environments are created a long time apart.
To me, it is not clear to what degree compiled bits are an integral part of the Julia depots, but this is my main problem with `~/.julia` currently (i.e. conda-forge Julia <=1.6).

My second concern is that it is not completely clear to me to what degree this PR really addresses the issues it closes. For example, #14 is about packaging Julia packages for conda-forge, i.e. how should I create the conda-forge feedstock `julia-rainfarm` from https://github.com/jhardenberg/RainFARM.jl. While sorting out the depot paths certainly helps and might be necessary, from reading this PR I still don't know how to do this or where to look up the documentation. In terms of documentation, at a minimum I would expect a bullet point in the [enumeration for new packages](https://conda-forge.org/docs/maintainer/adding_pkgs.html#getting-started).
This is probably beyond the scope of this PR, but in that case we might want to keep #14 open for now.

ngam commented:
> But if you want a bit more community engagement you might want to consider slowing down just a little bit. This PR, which closes 7 issues, some of which have been around for a long time and have been opened and worked upon by other people, took all of 5 days from creation to merging. Over the Christmas days.

Yeah, sorry... But, I'd rather interested people (like yourself) come back and reopen the issues and explain their current needs/expectations. It's very difficult for us to guess what people were thinking about 4+ years ago especially as the ecosystems (julia as well as conda) have moved so much since then. Obviously, we couldn't address all concerns (and you bring up really important ones) --- to help us push forward, could you please reopen issues/PRs that you think weren't adequately addressed and add your perspective?

I will start by opening two issues based on your two main concerns and we will work --- together and less rapidly --- to address them. 

> To me, it is not clear to what degree compiled bits are an integral part of the Julia depots, but this is my main problem with ~/.julia currently (i.e. conda-forge Julia <=1.6).

For this, @mkitti was of the opinion that we shouldn't change anything in the 1.6.x and just bring all newer changes to 1.7.1+. I concurred with his view (we don't want to mess up the setups of people who got comfortable with those versions...)

ngam commented:
> My second concern is that it is not completely clear to me to what degree this PR really addresses the issues it closes. For example, #14 is about packaging Julia packages for conda-forge, i.e. how should I create the conda-forge feedstock `julia-rainfarm` from https://github.com/jhardenberg/RainFARM.jl

I certainly didn't read it this way. I am not sure if this is even applicable now that julia has their own ecosystem and package-building setup. Our real goal here was to better isolate julia inside conda envs, so that a user can have many different julias (via conda) and also a separate standalone julia smoothly. How the user ends up using such isolation is obviously none of our business, right? And I am not sure if we should pursue changes to actually enmesh julia and conda more tightly; again, my thinking/perspective was to better isolate them for each other (not vice versa). But that's just one single opinion and I am happy to discuss this further and work on bringing changes/improvements you'd like to see :)

ngam commented:
> 7 issues

only 3. The others are PRs.

mkitti commented:
Native machine code is not cached by Julia during normal operation, so I do not expect there to be be ABI issues for compiled code generated from Julia unless one is making direct a`ccall`s. What is cached is lowered code and type inference on a per Julia environment basis in `$JULIA_DEPOT/compiled`. That said this does change between Julia versions and is under active development.

Thr chosen defaults for the 1.7 releases are quite conservative with unique Julia depots and environments for each conda environment. My concern is actually that is too conservative and that we are duplicating too many things.

The use of PackageCompiler.jl, which actually involves native code generation and storage is another matter, but its location of the generated system image shared library is determined by the user.

zklaus commented:
Thanks for your replies, @ngam and @mkitti. I'll pick up the discussion in #14, #164 and possibly new issues as needed.
